### DISCUSSION TITLE
Keep schedules after restart

### DISCUSSION THREAD

**Discussion by kuuroo3** [directus/directus#12851] [Archive]
Hi guys, 
I am having some trouble with my schedules..

I am selfhosting Directus with Docker and I need to restart Directus sometimes (at least for backupping the server).
It seems like schedules get destroyed whenever I restart.

Is there a way to keep my schedules? 

```
export default defineHook(({action, schedule}, {services, exceptions, logger, env}) => {

   action('mycollection.items.create', async (item, {schema}) => {
      schedule(cron_string(now, 3, env.PRODUCTION_MODE), async () => {
         // things my schedule does
      });
    });
 });
````

`cron_string` is a method which evaluates the string depending on the input and prod/local mode and shouldn't be of much interest.

**Comment by rijkvanzanten** [comment]
Right now, you're creating the `schedule` when the `action` fires. Don't you just want to have the schedule run regardless of that action execution?

```js
export default defineHook(({action, schedule}, {services, exceptions, logger, env}) => {
    schedule(cron_string(now, 3, env.PRODUCTION_MODE), async () => {
       // things my schedule does
    });
 });
```

**Reply by kuuroo3** [reply]
Yes and no... My usecase is like this: 
1. item created
1.1 instant info mail to multiple persons
1.2 start schedule for in 3 and 5 days -> Depending on the timestamp of the creation
1.2.1 if item wasn't edited yet send more mails

Therefore the schedule gets fired inside the action. If there is another way I gladly do that.

**Reply by rijkvanzanten** [reply]
Ah gotcha. In that case, there isn't really a way around the problem. The schedule isn't stored in the database / re-initialized on startup, so once the server resets, the original schedule start is lost. You could work around that by instead saving a record to the database of when the schedule should fire, and then using a schedule that fires every 10 minutes or so to check if there's any emails to send out for that time window 👍🏻

**Reply by kuuroo3** [reply]
Thanks for the idea. I will go ahead and try that.

If taking your snippet - would that make a difference in the scenario (concerning the restart)?
Would it be possible to change that behaviour in the future to keep all the schedules?

**Reply by rijkvanzanten** [reply]
> If taking your snippet - would that make a difference in the scenario (concerning the restart)?

Yeah, cause at that point, you're not relying on the registered schedule to keep track of the time when to fire. Because the schedule will check against the stored records in the database, it'll survive restarts etc as well (and it would allow you to cancel out the email early if that use case would ever arise) 🙂
